### DISCUSSION TITLE
ERROR	MongooseError: Operation `users.findOne()` buffering timed out after 10000ms

### DISCUSSION THREAD

**Discussion by m0x61h0x64i** [vercel/community#1016] [Help] [answered]
hi, i have pro vercel, and i am using nodejs & express & mongoose & mongodb to connect to the mongodb atlas database, after hundreds of deployments on hobby vercel account i didnt have database connecttion failure error like when i am using pro vercel! now when i deploy on pro vercel, my application fails to connect to the mongodb atlas database, however some deployments it connects fast and well but some other deployments it takes more than 5 minutes to connect the application to the mongodb atlas database! (in my vercel's hobby account it was okay and maybe it could connect in less than a minute!)

my connection code is the following:
```
const mongoose = require('mongoose')

const connectDatabase = async () => {
    try {      
        await mongoose.connect(process.env.MONGODB_URI)
        console.log('Connected to mongodb database successfully!')
    } catch (err) {
        console.log(err)
    }
}
connectDatabase()
```
mongodb version : `5.0.13`
mongoose version: `6.5.4`
vercel account type: `pro`

the error i get when i try to send a query using my deployed application on vercel to use mongodb atlas database after new deployments: 
`ERROR MongooseError: Operation users.findOne() buffering timed out after 10000ms`

had this error on hobby vercel account?
`true, but it took small amount of time to connect to the mongodb atlas database, some deployments connects fast and well.`

did you add your server ip address to mongodb atlas database > security > network aceess > ip address?
`yes, i allowed access from everywhere, 0.0.0.0/0`

Is your url correct in your connection string?
`yes, it sometimes connects fast and well.`

Is your mongoose.connect(......) code loading?
`yes, i imported database connection code on the top of the project codes to be executed first.`

Is your code inside a file? like config.js?
`i tried both ways, i imported my connection code inside and outside of index.js file.`

**Comment by amyegan** [comment, marked as answer]
Hi @m0x31h0x64i! Are you using the [MongoDB Atlas integration](https://vercel.com/integrations/mongodbatlas), or have you set up the connection manually? Timeouts often happen when a database is failing to connect. The first thing to check is that your [Environment Variables](https://vercel.com/docs/concepts/projects/environment-variables#) are present and not using expired tokens.

If your env vars are not the problem, this Stack Overflow post has a lot of good suggestions to try: https://stackoverflow.com/questions/65408618/mongooseerror-operation-users-findone-buffering-timed-out-after-10000ms

**Reply by amyegan** [reply]
If you can share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), I'm happy to jump in and help with debugging. Otherwise, it's difficult to determine what is going wrong without seeing it happen.

**Reply by m0x61h0x64i** [reply]
> If you can share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), I'm happy to jump in and help with debugging. Otherwise, it's difficult to determine what is going wrong without seeing it happen.

sure,

here is the link of the code to reproduce the issue:
https://codesandbox.io/s/stupefied-pike-lh0zlz?file=/src/index.js

1. deploy this app on the vercel.
2. click on the last deployment > functions > realtime logs.
3. visit the main page of the deployed app.
4. you will see errors related to connecting to the database on the logs (vercel applications have connection issue to the database).

the connection errors are temporary but it will be a cancer on pro plan it will cost  around 10-30 minutes to connect an app to the database!

**Comment by m0x61h0x64i** [comment]
hello @amyegan, thanks for your replay.

> Are you using the [MongoDB Atlas integration](https://vercel.com/integrations/mongodbatlas), or have you set up the connection manually?

im not using MongoDB Atlas integration, i have set up the connection manually in different ways and i tested different database providors(recently) but no fix.

> Environment Variables are present?

yes.

> this Stack Overflow post has a lot of good suggestions to try

I plowed google for this problem before.

**Reply by x3daniking** [reply]
a bit late but have solved the issue yet? getting the same issue dk what's wrong! i can access the collection through index.js but in my controller file, connection times out!

**Reply by ashishjullia** [reply]
@x3daniking any solution so far?

**Reply by x3daniking** [reply]
@ashishjullia yea, i've had fixed it. created a session and exported it. used only that session to query it and it worked somehow. 
If still doesn't work with you, try storing that session in a global variable upon initialization and use from that global variable. will work!

#while creating a client
this.client = new MongoClient(url);
global.session = this.client.startSession();

#in index.js
 await sessionManager.connect();

#while querying data
await global.session.client.db(db).collection(collection).findOne()
